controller/src/controller/subsystems/instrument_comm.py

```diff
@@ -300,49 +309,112 @@ async def _catch_expired_command(self) -> None:
+                        packet_type = SerialCommPacketTypes.STOP_STIM
+                    case {"command": "start_firmware_update"}:
+                        await self._handle_firmware_update(comm_from_monitor)
+                    case {
+                        "command": "trigger_firmware_error",
+                        "first_two_status_codes": first_two_status_codes,
+                    }:  # pragma: no cover
+                        packet_type = SerialCommPacketTypes.TRIGGER_ERROR
+                        bytes_to_send = bytes(first_two_status_codes)
+                    case {"command": "init_offline_mode"}:
+                        packet_type = SerialCommPacketTypes.INIT_OFFLINE_MODE
+                        self._system_in_offline_mode = True
+                    case {"command": "end_offline_mode"}:
+                        packet_type = SerialCommPacketTypes.END_OFFLINE_MODE
+                        # the _offline_state_change event needs to be triggered here instead of in process instrument comm because we first need to restart the task that handles instrument comm
+                        self._system_in_offline_mode = False
+                        self._offline_state_change.set()
+                    case {"command": "check_connection_status"}:
+                        packet_type = SerialCommPacketTypes.CHECK_CONNECTION_STATUS
+                    case invalid_comm:
+                        raise NotImplementedError(
+                            f"InstrumentComm received invalid comm from SystemMonitor: {invalid_comm}"
+                        )
+            else:
+                logger.info(
+                    f"Ignoring incoming command '{comm_from_monitor['command']}' in instrument comm while offline"
+                )
 
             if packet_type is not None:
                 await self._send_data_packet(packet_type, bytes_to_send)
                 await self._command_tracker.add(packet_type, comm_from_monitor)
 
+    async def _manage_online_mode_tasks(self) -> None:
+        main_task_name = self._wait_for_offline_state_change.__name__
+
+        # TODO clean up repetitive code
+        pending = {
+            asyncio.create_task(self._wait_for_offline_state_change(), name=main_task_name),
+            asyncio.create_task(self._handle_sending_handshakes()),
+            asyncio.create_task(self._handle_data_stream()),
+            asyncio.create_task(self._handle_beacon_tracking()),
+            asyncio.create_task(self._catch_expired_command()),
+        }
+
+        while True:
+            done, pending = await asyncio.wait(pending, return_when=asyncio.FIRST_COMPLETED)
+
+            for task in done:
+                task_name = task.get_name()
+
+                if task_name == main_task_name:
+                    if self._system_in_offline_mode:
+                        logger.info("Cancelling online mode tasks")
+                        exc = await clean_up_tasks(pending, ERROR_MSG)
```

Comment:
Wondering if it might make sense to making this a function. 
```
exc = await clean_up_tasks(pending, ERROR_MSG)
if exc:
   raise exc
```
Not sure if it's worth it yet though, so for now we can probably just add a comment